File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    N = len(wires)
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    
    def union(a, b):
        a = find(a); b = find(b)
        if a < b: parent[b] = a
        else: parent[a] = b
            
    diff = []
    for i in range(N):
        parent = [_ for _ in range(n+1)]
        for j in range(N):
            if i == j: continue
            union(wires[j][0], wires[j][1])
        
        for j in range(1, n+1): find(j)
        
        parent_nodes = list(set(parent[1:]))
        diff.append(abs(parent.count(parent_nodes[0]) - parent.count(parent_nodes[1])))
        
    return min(diff)

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

vector<int> parent;

int Find(int x)
{
    if (parent[x] != x)
    {
        parent[x] = Find(parent[x]);
    }
    return parent[x];
}

void Union(int a, int b)
{
    a = Find(a); b = Find(b);
    if (a < b)
    {
        parent[b] = a;
    }
    else
    {
        parent[a] = b;
    }
}

int solution(int n, vector<vector<int>> wires) {
    
    vector<int> diff;
    for (int i = 0; i < wires.size(); ++i)
    {
        parent.clear();
        for (int j = 0; j <= n; ++j)
        {
            parent.push_back(j);
        }
        
        for (int j = 0; j < wires.size(); ++j)
        {
            if (i == j)
            {
                continue;
            }
            Union(wires[j][0], wires[j][1]);
        }
        
        for (int j = 1; j <= n; ++j)
        {
            Find(j);
        }
        
        set<int> parent_nodes_set(parent.begin() + 1, parent.end());
        vector<int> parent_nodes_vec(parent_nodes_set.begin(), parent_nodes_set.end());
        
        diff.push_back(abs(count(parent.begin(), parent.end(), parent_nodes_vec[0]) - count(parent.begin(), parent.end(), parent_nodes_vec[1])));
    }
    
    return *min_element(diff.begin(), diff.end());
}